src/test/common_url_tests.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://opensource.org/license/mit/.
+
+#include <common/url.h>
+
+#include <string>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_SUITE(common_url_tests)
+
+// These test vectors were ported from test/regress.c in the libevent library
+// which used to be a dependency of the UrlDecode function.
+
+BOOST_AUTO_TEST_CASE(encode_decode_test) {
+    BOOST_CHECK_EQUAL(UrlDecode("Hello"), "Hello");
+    BOOST_CHECK_EQUAL(UrlDecode("99"), "99");
+    BOOST_CHECK_EQUAL(UrlDecode(""), "");
+    BOOST_CHECK_EQUAL(UrlDecode("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ123456789-.~_"),
+                      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ123456789-.~_");
+    BOOST_CHECK_EQUAL(UrlDecode("%20"), " ");
+    BOOST_CHECK_EQUAL(UrlDecode("%FF%F0%E0"), "\xff\xf0\xe0");
+    BOOST_CHECK_EQUAL(UrlDecode("%01%19"), "\x01\x19");
+    BOOST_CHECK_EQUAL(UrlDecode("http%3A%2F%2Fwww.ietf.org%2Frfc%2Frfc3986.txt"),
+                      "http://www.ietf.org/rfc/rfc3986.txt");
+    BOOST_CHECK_EQUAL(UrlDecode("1%2B2%3D3"), "1+2=3");
+}
+
+BOOST_AUTO_TEST_CASE(decode_malformed_test) {
+    BOOST_CHECK_EQUAL(UrlDecode("%%xhello th+ere \xff"), "%%xhello th+ere \xff");
+
+    BOOST_CHECK_EQUAL(UrlDecode("%"), "%");
```

Comment:
Hm, ok, that's another behavior change I didn't anticipate. Let me first describe my understanding of happens: In [the libevent version](https://github.com/libevent/libevent/blob/e0a4574ba2cbcdb64bb2b593e72be7f7f4010746/http.c#L3492) each of the two characters following the `%` are [looked at individually if they are valid hex characters](https://github.com/libevent/libevent/blob/e0a4574ba2cbcdb64bb2b593e72be7f7f4010746/http.c#L3508). The space is not so the if fails and the fallback case takes over, which just adds the `%` to res and then the next two chars are processed individually and since they are not special cases by themselves, they are also just added to res, so the result is `%1 `. In the current version here we look at both characters together with `std::from_chars`. The white space seems to be ignored, no error raised, and so the `decoded_value` is integer `1`. Then the `static_cast<char>(decoded_value)` converts the `1` to a non-printable character and so the result visually appears to be an empty string. It seems to me now that it's safer to check both digits individually if they are hex, so I have changed the code in this regard and added a few more tests with spaces. Let me know if you found other cases where the behavior is still changing.

Reply:
There's still one sneaky difference that I've found: `%00`.
Previously this prematurely terminated the string, now it's kept:
```C++
BOOST_CHECK_EQUAL(urlDecode("a%00b"), std::string("a\u0000b", 3));
```
this failed previously, passes now.

Reply:
Yes https://github.com/bitcoin/bitcoin/pull/29904#discussion_r1572090699

Reply:
Those cases were added since, I see, thanks.